### DISCUSSION TITLE
Using Rhino with kotlin and coroutines

### DISCUSSION THREAD

**Discussion by karel-zabloudil-3pg** [mozilla/rhino#1596] [General]
Is it possible to use Rhino with kotlin and coroutines ? Here is my use case. I want to run multiple scripts simultaneously using coroutines. Each script could call java/kotlin code which would read data from kotlin channel which could eventually suspend the coroutine. Due to nature of coroutines, the thread which run the suspended coroutine could be used to run another coroutine and another resumed script. That thread could be even completely different from the one which started the script.

Is this something which can be done with Rhino ? I'm afraid of possible synchronization and deadlock issues.

**Comment by tonygermano** [comment]
I'm not very familiar with how they work, but Rhino itself supports continuations to stop a script mid-way through and resume it later.

I'm not very familiar with kotlin coroutines either, but I would venture to guess that they would not suspend and switch contexts in the middle of executing a rhino script as executing a script would be a single statement from kotlin's point-of-view.

**Reply by karel-zabloudil-3pg** [reply]
How can you stop and resume script ?

**Reply by tonygermano** [reply]
Again, I'm not very familiar with them and don't use them myself, but here is a test for the api. https://github.com/mozilla/rhino/blob/650ec2991d95885cfafb149ce1b85e38bed2b5e6/rhino/src/test/java/org/mozilla/javascript/tests/ContinuationsApiTest.java

And here is the announcement in the release notes. It's actually a fairly old feature (~15 years old) https://rhino.github.io/releases/new_in_rhino_1.7r2.html#java-api-for-continuations

**Reply by tonygermano** [reply]
You do need to explicitly set your suspend points in the javascript code. It can't just break anywhere, so it won't work with arbitrary scripts. If you need to run arbitrary long-running scripts concurrently, you probably want to use real java threads as opposed to coroutines.

**Reply by karel-zabloudil-3pg** [reply]
Perferct, I think this is what I need.
Thanks.
